lib/trace/index.js

```diff
@@ -0,0 +1,156 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+'use strict';
+
+var errors = require('../errors');
+var safeParse = require('../util').safeParse;
+var toCamelCase = require('../util').toCamelCase;
+
+var core = require('./core');
+var TChannelTracer = require('./tchannel');
+var LogTracer = require('./log');
+
+function registerTraceEndpoints(ringpop) {
+    // XXX testing only?
+    //
+    // @wolski, we chatted briefly about this -- whatcha think should
+    // be done? the testing framework doesn't work, as is. (I can add
+    // more mocks to make it go.)
+    if (!ringpop || !ringpop.channel || !ringpop.membership) {
+        return;
+    }
+
+    registerTraceCommand(ringpop, '/trace/add', createTracer);
+    registerTraceCommand(ringpop, '/trace/remove', destroyTracer);
+}
+
+function registerTraceCommand(ringpop, endpoint, processor) {
+    var logPrefix = 'ringpop' + endpoint.replace(/\//g, '.');
+
+    ringpop.channel.register(endpoint, function onRequest(req, resp, optsStr) {
+        resp.headers.as = 'raw';
+        var opts = toCamelCase(safeParse(optsStr));
+        var config = core.resolveTracerConfig(ringpop, opts.event);
+        if (!config) {
+            var err = 'invalid event: ' + opts.event;
+            ringpop.logger.info(logPrefix + ' error: ' + err);
+            resp.sendNotOk(err);
+            return;
+        }
+
+        processor(ringpop, config, opts, function done(err) {
+            if (err) {
+                ringpop.logger.info(logPrefix + ' error: ' + err);
+                resp.sendNotOk(err);
+                return;
+            } 
+            ringpop.logger.info(logPrefix + ' success: ' + optsStr);
+            resp.sendOk();
+        });
+    });
+}
+
+function createTracer(ringpop, config, opts, callback) {
+    callback = callback || function() {};
+
+    if (!core.Tracer.validate(opts)) {
+        throw errors.InvalidOptionError({ option: 'sink.type', reason: 'is empty or unknown' });
+    }
+
+    if (indexOfTracer(ringpop, config, opts) >= 0) {
+        // idempotent, happy noop
+        callback();
+        return;
+    }
+
+    try {
+        var tracer = null;
+        switch (opts.sink.type) {
+        case 'tchannel':
+            tracer = new TChannelTracer(ringpop, config, opts);
+            break;
+
+        case 'log':
+            tracer = new LogTracer(ringpop, config, opts);
+            break;
+
+        default:
+            // can't occur since validate throws first, but jic.
```

Comment:
this is a side effect of doing type-specific validation within the constructors. I could do it explicitly instead, and then use traditional flow control. (I wrote it that way first, then went the other way.) Mostly a question of "good practice", and I don't have a notion on this particular issue. You?